### ISSUE TITLE
t.web2.channel.http.HTTPParser deals with HEAD requests incorrectly

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2302]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)| @ghazel reported|
|-|-|
|Trac ID|trac#2302|
|Type|defect|
|Created|2006-12-18 03:21:03Z|

HTTPParser.setConnectionParams has a comment:

```
                # If no Content-Length, indeterminate length data
                # (unless the responsecode was one of the special ones, or
                #  the request method was HEAD.
                # If the request was HEAD, self.length has been set to 0 by
                # HTTPClientRequest.submit)
                if self.code in http.NO_BODY_CODES:
                    self.length = 0
                else:
                    self.length = connHeaders.getHeader('content-length', self.length)
```

However, if the HEAD response has a content-length, it then sets self.length anyway. This leads to a pipelining error, as HTTPParser.lineReceived expects there to be more data.
 

<details><summary>Searchable metadata</summary>

```
trac-id__2302 2302
type__defect defect
reporter__ghazel ghazel
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__ 
time__1166412063000000 1166412063000000
changetime__1167694890000000 1167694890000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)<a name="note_1"></a>|@ghazel commented|
|-|-|

Also, simply setting stream.length to 0 if the method is HEAD does cause pipelining to work, but then as a user there's no way to tell what the content-length was. Very similar to this problem: [#2284](https://github.com/twisted/twisted/issues/2284)

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_2"></a>|@jyknight commented|
|-|-|

The patch on [#2284](https://github.com/twisted/twisted/issues/2284) fix for the bug, but I'm not quite sure where to put the content-length of the HEAD request. To be consistent, stream.length has to be 0.

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)<a name="note_3"></a>|@ghazel commented|
|-|-|

Here's the hack I threw together when I encountered these problems. It just stores content-length on a member variable which I access later

```
from twisted.web2.channel.http import HTTPParser

# http://twistedmatrix.com/trac/ticket/2302
old_setConnectionParams = HTTPParser.setConnectionParams
def setConnectionParams(self, connHeaders):
    old_setConnectionParams(self, connHeaders)
    self.content_length = self.length
    if hasattr(self, 'request'):
        if self.request.method == "HEAD":
            self.length = 0
HTTPParser.setConnectionParams = setConnectionParams


from twisted.web2.client.http import HTTPClientChannelRequest

# http://twistedmatrix.com/trac/ticket/2284
old_createRequest = HTTPClientChannelRequest.createRequest
def createRequest(self):
    old_createRequest(self)
    self.stream.length = self.length
    self.response.content_length = self.content_length
HTTPClientChannelRequest.createRequest = createRequest

# http://twistedmatrix.com/trac/ticket/2304
old_connectionLost = HTTPClientChannelRequest.connectionLost
def connectionLost(self, reason):
    self._error(reason)
HTTPClientChannelRequest.connectionLost = connectionLost
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [19162]) Revert r19155 - test suite regression

Refs [#2302](https://github.com/twisted/twisted/issues/2302)
Refs [#1870](https://github.com/twisted/twisted/issues/1870)

twisted.trial.test.test_runner.DeferredSharedSuiteTest.test_waitForDeferred fails with
gtk reactor.

twisted.trial.test.test_runner.TestSharedClassSuite.test_reporting fails with Python 2.5.

Some interaction seems to be taking place with the TODO feature.  testTLS and testBackwardsTLS
from twisted.test.test_ssl.SpammyTLSTestCase are reported as [FAIL] even though they are
marked TODO (only in some configurations).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

__no__, that Refs [#2303](https://github.com/twisted/twisted/issues/2303). :/

**Comment by twisted-trac** [comment]
|<img alt="foom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|foom set status to `closed`|
|-|-|

(In [19180]) Fix content-length issues in HTTP client:
Propagate content-length into stream.length. (Fixes [#2284](https://github.com/twisted/twisted/issues/2284))
Fix HEAD requests with content-length. (Fixes [#2302](https://github.com/twisted/twisted/issues/2302))
Also, refactor and fix the test_client tests and add some docstrings.

Reporter: ghazel
Author: jknight
Reviewer: jerub, exarkun

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
